### ISSUE TITLE
[Proposal]: Reorganize context type.

### ISSUE THREAD

**Issue by mike820324** [mike820324/microProxy#87]
### Introduction:

Currently, we have two message context type in this project.
- `context.py`: used by all the layers.
- `http.py`: used by viewer.

However, as the project becomes bigger, there may be more context types. Therefore, I think we should organize them in a much better structure.
### Proposal:

We can categorize the context types into three different categories which are,
- layer context: the context type that is used by different layer and `LayerManager`.
- viewer context: the context type that will be used to communicate with viewer.
- plugin context: the context type that will be used to communicate with plug script.

I propose that we can organize these context into the following folder structure.
- microproxy
  - context
    - `layer.py`: Implement `LayerContextBuilder` class.
    - `viewer.py`: Implement `ViewerContextBuilder` class
    - `plugin.py`: mImplement `PluginContextBuilder` class.
    - `http.py`: used by both viewer.py and plugin.py.

Each context type should have at least the following methods.
- `copy`: which will copy the data using `copy.copy`.
- `serialize`: change the context into a json compatible format.
- `deserialize`: restore the json compatible format into object.
#### HttpRequest:

Represent the Http Request type for both http1 or http2. There are five fields which are.
- timestamp: unix timestamp when the request comes in.
- version: http version. [http1.1/ h2]
- method: http method.
- path: path of this connection
- headers: http headers
- body: a base64 encoded http body raw data.

The following is a simple example 

``` json
{
    "timestamp": 1467349643,
    "version": "http1.1",
    "method": "GET",
    "path": "/hello.html",
    "headers": {"Host": "google.com"},
    "body": "base64 encoded content"
}
```
#### HttpResponse

Represent the Http Response type for both http1 or http2. There are five fields which are.
- timestamp: unix timestamp when the response comes in.
- version: http version. [http1.1/ h2]
- code: http response code.
- reason: http response status reason.
- headers: http headers
- body: a base64 encoded http body raw data.

The following is a simple example 

``` json
{
    "timestamp": 1467349665,
    "version": "http1.1",
    "code": "200",
    "reason": "OK",
    "headers": {"Host": "google.com"},
    "body": "base64 encoded content"
}
```
#### ViewerContext:

``` json
{
    "scheme": "h2",
    "host": "www.google.com",
    "port": 443,
    "path": "/hello_world",
    "request": {
         "timestamp": 1467349665,
         "version": "http1.1",
         "code": "200",
         "reason": "OK",
         "headers": {"Host": "google.com"},
         "body": "base64 encoded content"
    },
    "response": {
        "timestamp": 1467349665,
        "version": "http1.1",
        "code": "200",
        "reason": "OK",
        "headers": {"Host": "google.com"},
        "body": "base64 encoded content"
    }
}
```
#### PluginContext:

``` json
{
    "scheme": "h2",
    "host": "www.google.com",
    "port": 443,
    "path": "/hello_world",
    "request": {
         "timestamp": 1467349665,
         "version": "http1.1",
         "code": "200",
         "reason": "OK",
         "headers": {"Host": "google.com"},
         "body": "base64 encoded content"
    },
    "response": {
        "timestamp": 1467349665,
        "version": "http1.1",
        "code": "200",
        "reason": "OK",
        "headers": {"Host": "google.com"},
        "body": "base64 encoded content"
    }
}
```

**Comment by chhsiao90** [comment]
Nice suggestion!

**Comment by chhsiao90** [comment]
We should work on this ASAP!
Maybe we should not process on new issue but close the in processing issue and process on this one!

**Comment by mike820324** [comment]
Will work on it in this weekend.

**Comment by mike820324** [comment]
Closed via PR #103.
